File: Chain of Responsibility/ChainOfResponsibility.cpp
#include <iostream>
using namespace std;

// Handler 
class Handler{
public:
    Handler* nextHandler;
    virtual void handle() = 0;
    virtual void setNextHandler(Handler* handler) = 0;
};

// Concrete handlers
class Manager: public Handler{
public:
    Handler* nextHandler;
    void handle()
    {
        cout<<"Request was handled by Manager"<<endl;
        if(nextHandler)
            nextHandler->handle();
    }
    void setNextHandler(Handler* handler)
    {
        nextHandler = handler;
    }
};

class HR: public Handler{
public:
    Handler* nextHandler;
    void handle()
    {
        cout<<"Request was handled by HR"<<endl;
        if(nextHandler)
            nextHandler->handle();
    }
    void setNextHandler(Handler* handler)
    {
        nextHandler = handler;
    }
};

class CTO: public Handler{
public:
    Handler* nextHandler;
    void handle()
    {
        cout<<"Request was handled by CTO"<<endl;
        if(nextHandler)
            nextHandler->handle();
    }
    void setNextHandler(Handler* handler)
    {
        nextHandler = handler;
    }
};

int main() {
    Handler* cto = new CTO();
    Handler* hr = new HR();
    Handler* manager = new Manager();
    manager->setNextHandler(hr);
    hr->setNextHandler(cto);
    cto->setNextHandler(NULL);
    manager->handle();
}

File: Chain of Responsibility/ChainOfResponsibility.java
import java.util.*;

interface Handler {
    public void setNextHandler(Handler handler);
    public void handle();
}

class HR implements Handler {
    private Handler nextHandler;
    @Override
    public void setNextHandler(Handler handler)
    {
        this.nextHandler = handler;
    }
    @Override
    public void handle() {
        System.out.println("Processed by HR");
        if(this.nextHandler != null)
            this.nextHandler.handle();
    }
}
class Manager implements Handler {
    private Handler nextHandler;
    @Override
    public void setNextHandler(Handler handler)
    {
        this.nextHandler = handler;
    }
    @Override
    public void handle() {
        System.out.println("Processed by Manager");
        if(this.nextHandler != null)
            this.nextHandler.handle();
    }
}
class CEO implements Handler {
    private Handler nextHandler;
    @Override
    public void setNextHandler(Handler handler)
    {
        this.nextHandler = handler;
    }
    @Override
    public void handle() {
        System.out.println("Processed by CEO");
        if(this.nextHandler != null)
            this.nextHandler.handle();
    }
}
class Main {
    public static void main(String[] args)
    {
        Handler hr = new HR();
        Handler manager = new Manager();
        Handler ceo = new CEO();
        hr.setNextHandler(manager);
        manager.setNextHandler(ceo);
        hr.handle();
    }
}